slo-monitor/src/monitors/store.go

```diff
@@ -0,0 +1,207 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package monitors
+
+import (
+	"fmt"
+	"sync"
+
+	"k8s.io/client-go/tools/cache"
+
+	"github.com/golang/glog"
+)
+
+const (
+	setOperationType    = "set"
+	deleteOperationType = "delete"
+)
+
+type workItem struct {
+	operationType string
+	key           string
+	obj           interface{}
+}
+
+// NoStoreQueue implements k8s.io/client-go/tools/cache Store and Queue interfaces
+type NoStoreQueue struct {
+	sync.Mutex
+
+	cond sync.Cond
+	// queue of items to process
+	queue []workItem
+	// map from keys to indices of items to process. Indices are not modified and are always growing,
+	// (i.e. next item gets an index of the last item + 1). To map it to "real" indices inside the queue
+	// we also store the offset (index of last processed item + 1).
+	indices map[string]int64
+	// index that will be assigned to next item that'll be added to the queue
+	nextIndex int64
+	// index of the last process item + 1
+	offset int64
+}
+
+var keyFunc = cache.DeletionHandlingMetaNamespaceKeyFunc
+
+// NewNoStoreQueue returns newly created NoStoreQueue
+func NewNoStoreQueue() *NoStoreQueue {
+	q := &NoStoreQueue{
+		queue:     []workItem{},
+		indices:   map[string]int64{},
+		nextIndex: 0,
+		offset:    0,
+	}
+
+	q.cond.L = q
+	return q
+}
+
+func (q *NoStoreQueue) addOrModifyItem(key string, obj interface{}, operationType string) {
```

Comment:
It's used for all add/update/delete operations. So the name is pretty misleading.

Maybe "queueItemOperation" or sth like that?